lib/puppet/provider/ec2_instance/v2.rb

```diff
@@ -46,10 +46,11 @@ def self.instance_to_hash(region, instance)
     end
     subnet_name = nil
     if instance.subnet_id
-      subnet_response = ec2_client(region).describe_subnets(subnet_ids: [instance.subnet_id])
-      subnet_name_tag = subnet_response.data.subnets.first.tags.detect { |tag| tag.key == 'Name' }
+      subnet_response = ec2_client(region).describe_subnets(filters: [
+        {name: 'subnet-id', values: [instance.subnet_id]}
+      ])
+      subnet_name = subnet_response.data.subnets.empty? ? nil : name_from_tag(subnet_response.data.subnets.first)
```

Comment:
It's also worth noting we've not seen any failures of that type in the acceptance tests (we generally assert on attributes coming back in the puppet resource tests.

What we did see were issues related to state in AWS being misleading. ie. AWS returning a resource id in an association that has already been deleted. This is obviously an eventual consistency thing on the AWS side which makes perfect sense.

Note that not all AWS resources have constrains around deleting connected resources, although some do.

This all means we will regularly receive IDs from AWS that don't exist, and failing in that condition isn't the correct thing to do. Even post-Puppet 4 failing in all these cases would mean regularly seeing entire runs fail due to the valid but inconsistent state of AWS.